File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyTouchHelperCallback.java
package com.airbnb.epoxy;

import android.graphics.Canvas;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.ViewHolder;
import android.support.v7.widget.helper.ItemTouchHelper;

import java.util.List;

/**
 * A wrapper around {@link android.support.v7.widget.helper.ItemTouchHelper.Callback} to cast all
 * view holders to {@link com.airbnb.epoxy.EpoxyViewHolder} for simpler use with Epoxy.
 */
public abstract class EpoxyTouchHelperCallback extends ItemTouchHelper.Callback {

  @Override
  public final int getMovementFlags(RecyclerView recyclerView, ViewHolder viewHolder) {
    return getMovementFlags(recyclerView, (EpoxyViewHolder) viewHolder);
  }

  /**
   * @see #getMovementFlags(RecyclerView, ViewHolder)
   */
  protected abstract int getMovementFlags(RecyclerView recyclerView, EpoxyViewHolder viewHolder);

  @Override
  public final boolean onMove(RecyclerView recyclerView, ViewHolder viewHolder, ViewHolder target) {
    return onMove(recyclerView, (EpoxyViewHolder) viewHolder, (EpoxyViewHolder) target);
  }

  /**
   * @see #onMove(RecyclerView, ViewHolder, ViewHolder)
   */
  protected abstract boolean onMove(RecyclerView recyclerView, EpoxyViewHolder viewHolder,
      EpoxyViewHolder target);

  @Override
  public final void onSwiped(ViewHolder viewHolder, int direction) {
    onSwiped((EpoxyViewHolder) viewHolder, direction);
  }

  /**
   * @see #onSwiped(ViewHolder, int)
   */
  protected abstract void onSwiped(EpoxyViewHolder viewHolder, int direction);

  @Override
  public final boolean canDropOver(RecyclerView recyclerView, ViewHolder current,
      ViewHolder target) {
    return canDropOver(recyclerView, (EpoxyViewHolder) current, (EpoxyViewHolder) target);
  }

  /**
   * @see #canDropOver(RecyclerView, ViewHolder, ViewHolder)
   */
  protected boolean canDropOver(RecyclerView recyclerView, EpoxyViewHolder current,
      EpoxyViewHolder target) {
    return super.canDropOver(recyclerView, current, target);
  }

  @Override
  public final float getSwipeThreshold(ViewHolder viewHolder) {
    return getSwipeThreshold((EpoxyViewHolder) viewHolder);
  }

  /**
   * @see #getSwipeThreshold(ViewHolder)
   */
  protected float getSwipeThreshold(EpoxyViewHolder viewHolder) {
    return super.getSwipeThreshold(viewHolder);
  }

  @Override
  public final float getMoveThreshold(ViewHolder viewHolder) {
    return getMoveThreshold((EpoxyViewHolder) viewHolder);
  }

  /**
   * @see #getMoveThreshold(ViewHolder)
   */
  protected float getMoveThreshold(EpoxyViewHolder viewHolder) {
    return super.getMoveThreshold(viewHolder);
  }

  @Override
  public final ViewHolder chooseDropTarget(ViewHolder selected, List dropTargets, int curX,
      int curY) {
    //noinspection unchecked
    return chooseDropTarget((EpoxyViewHolder) selected, (List<EpoxyViewHolder>) dropTargets, curX,
        curY);
  }

  /**
   * @see #chooseDropTarget(ViewHolder, List, int, int)
   */
  protected EpoxyViewHolder chooseDropTarget(EpoxyViewHolder selected,
      List<EpoxyViewHolder> dropTargets, int curX, int curY) {

    //noinspection unchecked
    return (EpoxyViewHolder) super.chooseDropTarget(selected, (List) dropTargets, curX, curY);
  }

  @Override
  public final void onSelectedChanged(ViewHolder viewHolder, int actionState) {
    onSelectedChanged((EpoxyViewHolder) viewHolder, actionState);
  }

  /**
   * @see #onSelectedChanged(ViewHolder, int)
   */
  protected void onSelectedChanged(EpoxyViewHolder viewHolder, int actionState) {
    super.onSelectedChanged(viewHolder, actionState);
  }

  @Override
  public final void onMoved(RecyclerView recyclerView, ViewHolder viewHolder, int fromPos,
      ViewHolder target, int toPos, int x, int y) {

    onMoved(recyclerView, (EpoxyViewHolder) viewHolder, fromPos, (EpoxyViewHolder) target, toPos, x,
        y);
  }

  /**
   * @see #onMoved(RecyclerView, ViewHolder, int, ViewHolder, int, int, int)
   */
  protected void onMoved(RecyclerView recyclerView, EpoxyViewHolder viewHolder, int fromPos,
      EpoxyViewHolder target, int toPos, int x, int y) {
    super.onMoved(recyclerView, viewHolder, fromPos, target, toPos, x, y);
  }

  @Override
  public final void clearView(RecyclerView recyclerView, ViewHolder viewHolder) {
    clearView(recyclerView, (EpoxyViewHolder) viewHolder);
  }

  /**
   * @see #clearView(RecyclerView, ViewHolder)
   */
  protected void clearView(RecyclerView recyclerView, EpoxyViewHolder viewHolder) {
    super.clearView(recyclerView, viewHolder);
  }

  @Override
  public final void onChildDraw(Canvas c, RecyclerView recyclerView, ViewHolder viewHolder,
      float dX,
      float dY, int actionState, boolean isCurrentlyActive) {

    onChildDraw(c, recyclerView, (EpoxyViewHolder) viewHolder, dX, dY, actionState,
        isCurrentlyActive);
  }

  /**
   * @see #onChildDraw(Canvas, RecyclerView, ViewHolder, float, float, int, boolean)
   */
  protected void onChildDraw(Canvas c, RecyclerView recyclerView, EpoxyViewHolder viewHolder,
      float dX, float dY, int actionState, boolean isCurrentlyActive) {
    super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
  }

  @Override
  public final void onChildDrawOver(Canvas c, RecyclerView recyclerView, ViewHolder viewHolder,
      float dX,
      float dY, int actionState, boolean isCurrentlyActive) {

    onChildDrawOver(c, recyclerView, (EpoxyViewHolder) viewHolder, dX, dY, actionState,
        isCurrentlyActive);
  }

  /**
   * @see #onChildDrawOver(Canvas, RecyclerView, ViewHolder, float, float, int, boolean)
   */
  protected void onChildDrawOver(Canvas c, RecyclerView recyclerView, EpoxyViewHolder viewHolder,
      float dX, float dY, int actionState, boolean isCurrentlyActive) {

    super.onChildDrawOver(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
  }
}

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyTouchHelperCallback.kt
package com.airbnb.epoxy

import android.graphics.Canvas
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

/**
 * A wrapper around [androidx.recyclerview.widget.ItemTouchHelper.Callback] to cast all
 * view holders to [com.airbnb.epoxy.EpoxyViewHolder] for simpler use with Epoxy.
 */
abstract class EpoxyTouchHelperCallback : ItemTouchHelper.Callback() {

    override fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder
    ): Int = getMovementFlags(recyclerView, viewHolder as EpoxyViewHolder)

    /**
     * @see getMovementFlags
     */
    protected abstract fun getMovementFlags(
        recyclerView: RecyclerView,
        viewHolder: EpoxyViewHolder
    ): Int

    override fun onMove(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder
    ): Boolean = onMove(recyclerView, viewHolder as EpoxyViewHolder, target as EpoxyViewHolder)

    /**
     * @see onMove
     */
    protected abstract fun onMove(
        recyclerView: RecyclerView,
        viewHolder: EpoxyViewHolder,
        target: EpoxyViewHolder
    ): Boolean

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int): Unit =
        onSwiped(viewHolder as EpoxyViewHolder, direction)

    /**
     * @see onSwiped
     */
    protected abstract fun onSwiped(viewHolder: EpoxyViewHolder, direction: Int)

    override fun canDropOver(
        recyclerView: RecyclerView,
        current: RecyclerView.ViewHolder,
        target: RecyclerView.ViewHolder
    ): Boolean = canDropOver(recyclerView, current as EpoxyViewHolder, target as EpoxyViewHolder)

    /**
     * @see canDropOver
     */
    protected open fun canDropOver(
        recyclerView: RecyclerView,
        current: EpoxyViewHolder,
        target: EpoxyViewHolder
    ): Boolean = super.canDropOver(recyclerView, current, target)

    override fun getSwipeThreshold(viewHolder: RecyclerView.ViewHolder): Float =
        getSwipeThreshold(viewHolder as EpoxyViewHolder)

    /**
     * @see getSwipeThreshold
     */
    protected fun getSwipeThreshold(viewHolder: EpoxyViewHolder): Float =
        super.getSwipeThreshold(viewHolder)

    override fun getMoveThreshold(viewHolder: RecyclerView.ViewHolder): Float =
        getMoveThreshold(viewHolder as EpoxyViewHolder)

    /**
     * @see getMoveThreshold
     */
    protected fun getMoveThreshold(viewHolder: EpoxyViewHolder): Float =
        super.getMoveThreshold(viewHolder)

    @Suppress("UNCHECKED_CAST")
    override fun chooseDropTarget(
        selected: RecyclerView.ViewHolder,
        dropTargets: List<RecyclerView.ViewHolder>,
        curX: Int,
        curY: Int
    ): EpoxyViewHolder? = chooseDropTarget(
        selected as EpoxyViewHolder,
        dropTargets as List<EpoxyViewHolder>,
        curX,
        curY
    )

    /**
     * @see chooseDropTarget
     */
    protected fun chooseDropTarget(
        selected: EpoxyViewHolder,
        dropTargets: List<EpoxyViewHolder>,
        curX: Int,
        curY: Int
    ): EpoxyViewHolder? =
        super.chooseDropTarget(selected, dropTargets, curX, curY) as? EpoxyViewHolder

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int): Unit =
        onSelectedChanged(viewHolder as EpoxyViewHolder?, actionState)

    /**
     * @see onSelectedChanged
     */
    protected open fun onSelectedChanged(viewHolder: EpoxyViewHolder?, actionState: Int): Unit =
        super.onSelectedChanged(viewHolder, actionState)

    override fun onMoved(
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        fromPos: Int,
        target: RecyclerView.ViewHolder,
        toPos: Int,
        x: Int,
        y: Int
    ): Unit = onMoved(
        recyclerView,
        viewHolder as EpoxyViewHolder,
        fromPos,
        target as EpoxyViewHolder,
        toPos,
        x,
        y
    )

    /**
     * @see onMoved
     */
    protected fun onMoved(
        recyclerView: RecyclerView,
        viewHolder: EpoxyViewHolder,
        fromPos: Int,
        target: EpoxyViewHolder,
        toPos: Int,
        x: Int,
        y: Int
    ): Unit = super.onMoved(recyclerView, viewHolder, fromPos, target, toPos, x, y)

    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Unit =
        clearView(recyclerView, viewHolder as EpoxyViewHolder)

    /**
     * @see clearView
     */
    protected open fun clearView(recyclerView: RecyclerView, viewHolder: EpoxyViewHolder): Unit =
        super.clearView(recyclerView, viewHolder)

    override fun onChildDraw(
        c: Canvas,
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder,
        dX: Float,
        dY: Float,
        actionState: Int,
        isCurrentlyActive: Boolean
    ): Unit = onChildDraw(
        c,
        recyclerView,
        viewHolder as EpoxyViewHolder,
        dX,
        dY,
        actionState,
        isCurrentlyActive
    )

    /**
     * @see onChildDraw
     */
    protected open fun onChildDraw(
        c: Canvas,
        recyclerView: RecyclerView,
        viewHolder: EpoxyViewHolder,
        dX: Float,
        dY: Float,
        actionState: Int,
        isCurrentlyActive: Boolean
    ): Unit = super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)

    override fun onChildDrawOver(
        c: Canvas,
        recyclerView: RecyclerView,
        viewHolder: RecyclerView.ViewHolder?,
        dX: Float,
        dY: Float,
        actionState: Int,
        isCurrentlyActive: Boolean
    ) {
        onChildDrawOver(
            c, recyclerView, viewHolder as? EpoxyViewHolder, dX, dY, actionState,
            isCurrentlyActive
        )
    }

    /**
     * @see onChildDrawOver
     */
    protected fun onChildDrawOver(
        c: Canvas,
        recyclerView: RecyclerView,
        viewHolder: EpoxyViewHolder?,
        dX: Float,
        dY: Float,
        actionState: Int,
        isCurrentlyActive: Boolean
    ): Unit = super.onChildDrawOver(
        c,
        recyclerView,
        viewHolder,
        dX,
        dY,
        actionState,
        isCurrentlyActive
    )
}
